DOC: Versioning, public/private API, deprecation (#10096)
spacetelescope/jwst: docs/jwst/user_documentation/more_information.rst

.. _jwst-version-scheme:

Versioning Scheme
-----------------

.. note::

   This section addresses the release version to PyPI.
   It is not to be confused with the DMS quarterly build version.

The ``jwst`` package does not follow strict `semantic versioning <https://semver.org/>`_, in that a "minor" version bump may also contain non-backward compatible changes. However, a "patch" release will strive to contain only bug fixes.

.. _jwst-public-vs-private-api:

API: Public vs Private
----------------------

As per Python convention, any API name that starts with underscore (e.g., ``_my_private_function``) is considered private; i.e., **it could be removed and changed without notice.** Any API not officially documented (i.e., you only found it after some extensive code-diving) is also considered private.

However, contrary to common practice, ``jwst`` package also considers what would normally be public API (i.e., no leading underscore and fully documented) to be private unless stated below.

These are the only conditions where ``jwst`` would consider public API and thus follows :ref:`jwst-deprecation-policy`:

* High-level calibration step classes

> * Command-line interface for calibration steps

Reviewer: It'd be good to add some examples here, e.g. 
```suggestion
* Calibration step classes (e.g., Spec2Pipeline, JumpStep, AmiNormalizeStep)
* Command-line interface for calibration steps (e.g. calwebb_image3)
```

Reviewer: Are there more things to add here?

I'm not too familiar with the scripts inside our `scripts/` subdirectory.  Are we ok modifying the command-line interface to those without a deprecation period?

I know it's technically an stdatamodels thing, but should we add in something like "Changes that make files no longer readable as datamodels, e.g. removal of a WCS transform converter"?  Or perhaps that is too technical?

Reviewer 2: I also think it would be helpful to mention datamodel classes here.

I think the scripts we have are intended more for developer use than user support.  There's nothing in there that a regular pipeline user should be relying on.  I don't think they need a deprecation period.

Author: Added data models.

However, a "patch" release will strive to contain only bug fixes.

.. _jwst-public-vs-private-api:

API: Public vs Private
----------------------

As per Python convention, any API name that starts with underscore
(e.g., ``_my_private_function``) is considered private.
Any API not officially documented (i.e., you only found it after some extensive
code-diving) is also considered private.

However, contrary to common practice, ``jwst`` package also considers what would
normally be public API (i.e., no leading underscore and fully documented)
to be private unless stated as follows. **Any API not listed below, even if
fully documented, could be removed or changed without notice.**

These are the only conditions where ``jwst`` would consider public API and
thus follows :ref:`jwst-deprecation-policy`:

* Calibration step classes (e.g., Spec2Pipeline, JumpStep, AmiNormalizeStep)
* Command-line interface for calibration steps (e.g., calwebb_image3)
* Data models (e.g., removal of a WCS transform converter causing I/O incompatibility)

.. _jwst-deprecation-policy:

API Deprecation Policy
----------------------

.. note::

>    This section only cover API changes that are not backward-compatible.

Reviewer 2: ```suggestion
   This section only covers API changes that are not backward-compatible.
```

   Adding a new API is usually backward compatible and thus not covered here.

Occasionally, there is a need to modify or remove a given API
(e.g., introduction of new algorithm or instrumentation changes).
For public API (see :ref:`jwst-public-vs-private-api`), when such a
breaking change happens, if possible, ``jwst`` would first deprecate
the item being removed or renamed; this could be a input keyword,
step configuration, function, class, etc.

A deprecated item would emit a ``DeprecationWarning`` via Python warnings

> system (and also to log file, where appropriate). Ideally, the deprecation

Reviewer 2: ```suggestion
A deprecated item would emit a ``DeprecationWarning`` via the Python warnings
system (and also to the log, where appropriate). Ideally, the deprecation
```

warning would provide at least the following information:

> * which ``jwst`` version was it first deprecated, and

Reviewer 2: ```suggestion
* which ``jwst`` version first deprecated it, and
```
